Add Automation

nsatragno opened the pull request.

nsatragno committed:
Automation: AddVirtualAuthenticator

nsatragno committed:
Automation: RemoveVirtualAuthenticator.

nsatragno committed:
Automation: AddCredential

nsatragno committed:
More Automation commands.

nsatragno committed:
Remove attachment from API.

equalsJeffH committed:
merge from master and fix conflict

nsatragno committed:
Replace <a></a> with [==]

nsatragno committed:
Use correct syntax for authenticator ID URI param.

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
 :: None.
 
 
+# Automation # {#sctn-automation}
```
```suggestion
# User Agent Automation # {#sctn-automation}
```

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
 
+# Automation # {#sctn-automation}
+
+For the purposes of user-agent automation and application testing, this document defines a number of [=extension
```
```suggestion
For the purposes of user agent automation and [=web application=] testing, this document defines a number of [[WebDriver]] [=extension commands=].
```

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+# Automation # {#sctn-automation}
+
+For the purposes of user-agent automation and application testing, this document defines a number of [=extension
+commands=] for the [[WebDriver]] specification. 
```
```suggestion

```

equalsJeffH replied:
```suggestion
```

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+## <dfn>Virtual Authenticators</dfn> ## {#sctn-automation-virtual-authenticators}
+
+These [=extension commands=] create and interact with [=Virtual Authenticators=]: software implementations of the
+<a>Authenticator Model</a>. <a>Virtual Authenticators</a> are stored in a <dfn>Virtual Authenticator Database</dfn> and
```
```suggestion
<a>Authenticator Model</a>. <a>Virtual Authenticators</a> are stored in a <dfn>Virtual Authenticator Database</dfn>.
```

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+
+These [=extension commands=] create and interact with [=Virtual Authenticators=]: software implementations of the
+<a>Authenticator Model</a>. <a>Virtual Authenticators</a> are stored in a <dfn>Virtual Authenticator Database</dfn> and
+have the following properties:
```
```suggestion
Each stored [=virtual authenticator-] has the following properties:
```

equalsJeffH replied:
Also, in terms of virftAuthnr (virtual authenticator) properties, I wonder if we ought to also allow for user presence test?

nsatragno replied:
Applied the suggestion % the typo on the reference.

> Also, in terms of virftAuthnr (virtual authenticator) properties, I wonder if we ought to also allow for user presence test?

Yes, I completely forgot to add this >_<.  Added `isUserPresent`.

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+have the following properties:
+
+: |authenticatorId|
+:: An arbitrary non-null string that uniquely identifies the [=Virtual Authenticator=].
```
since this string will be embedded into URIs as a path segment, we ought to constrain its format, e.g., both its length, and allowable chars -- yes?
according to [RFC3986](https://tools.ietf.org/html/rfc3986) segments are made up of `pchar` which is defined  as `unreserved / pct-encoded / sub-delims / ":" / "@"`  --  I am thinking we ought to simply stipulate that `authenticatorId` is composed of `unreserved` chars and can be up to, say, 48 chars long. This would allow for using UUIDs as `authenticatorId`s it appears from [RFC4122](https://tools.ietf.org/html/rfc4122) that a UUID in string representation as a URN is 43 chars long.

note: WebDriver stipulates `session id`s are UUIDs

nsatragno replied:
Good point - updated the text here to constrain the string to be made of up to 48 unreserved characters.

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+: |authenticatorId|
+:: An arbitrary non-null string that uniquely identifies the [=Virtual Authenticator=].
+: |protocol|
+:: The protocol the [=Virtual Authenticator=] uses to encode its messages. Either `ctap2` for the [[FIDO-CTAP]] format
```
```suggestion
:: The protocol the [=Virtual Authenticator=] speaks: either `ctap2` or `ctap1-u2f` [[FIDO-CTAP]].
```

equalsJeffH replied:
Note that the CTAP spec, now that FIDO pub'd a "proposed standard" version, supersedes the old FIDO-U2F-Message-Formats spec. The former actually includes the latter.  FIDO needs to figure out how to mark old specs as deprecated...

Also, various actual authenticators support multiple CTAP protocol versions. Ought we allow for instantiating virftAuthnrs (virtual authenticators) that also do so?

nsatragno replied:
> Also, various actual authenticators support multiple CTAP protocol versions. Ought we allow for instantiating virftAuthnrs (virtual authenticators) that also do so?

We can certainly do this. Do you see any test case where it might be useful? I had doubts about adding the protocol version in the first place because it should be transparent to the client of the API...

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+:: An arbitrary non-null string that uniquely identifies the [=Virtual Authenticator=].
+: |protocol|
+:: The protocol the [=Virtual Authenticator=] uses to encode its messages. Either `ctap2` for the [[FIDO-CTAP]] format
+    or `u2f` for the [[FIDO-U2F-Message-Formats]].
```
```suggestion
```

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+:: The {{AuthenticatorTransport}} simulated. If the |transport| is set to {{AuthenticatorTransport/internal}}, the
+    authenticator simulates [=platform attachment=]. Otherwise, it simulates [=cross-platform attachment=].
+: |hasResidentKey|
+:: If set to `true`, the authenticator will support [=resident credentials=].
```
```suggestion
:: If set to [TRUE] the authenticator will support [=resident credentials=].
```

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+: |hasResidentKey|
+:: If set to `true`, the authenticator will support [=resident credentials=].
+: |hasUserVerification|
+:: If set to `true`, the authenticator will support [=user verification=].
```
```suggestion
:: If set to [TRUE], the authenticator supports [=user verification=].
```

equalsJeffH commented on index.bs:
```diff
@@ -228,6 +228,20 @@ spec: webidl; urlPrefix: https://heycam.github.io/webidl
+        text: WebDriver error; url: dfn-error
+        text: WebDriver error code; url: dfn-error-code
+        text: extension command; url: dfn-extension-command
+        text: extension command name; url: dfn-extension-command-name
```
```suggestion
        text: extension command name; url: dfn-extension-command-name
        text: getting a property; url: dfn-getting-properties
```

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+     [=invalid argument=].
+ 1. Let |protocol| be the result of [=trying=] to get the |parameter|'s {{AddVirtualAuthenticatorParameters/protocol}}
+    property.
+ 1. If |protocol| is null, return a [=WebDriver error=] with [=WebDriver error code=] [=invalid argument=].
```
```suggestion
```

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+
+ 1. If |parameters| is not a JSON [=Object=], return a [=WebDriver error=] with [=WebDriver error code=]
+     [=invalid argument=].
+ 1. Let |protocol| be the result of [=trying=] to get the |parameter|'s {{AddVirtualAuthenticatorParameters/protocol}}
```
```suggestion
 1. Let |protocol| be the result of [=getting a property=] named {{AddVirtualAuthenticatorParameters/protocol}} from the |parameters| argument, else return a [=WebDriver error=] with [=WebDriver error code=] [=invalid argument=].
```
from the webdriver spec it appears that one only uses the "trying" thing when calling a "fallible" algorithm, and "getting a property" is not one of them. e.g. see the [set window rect](https://www.w3.org/TR/webdriver/#set-window-rect) remote end steps.

nsatragno replied:
Done - I "refactored" this section taking this suggestion into account.

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+ 1. Let |protocol| be the result of [=trying=] to get the |parameter|'s {{AddVirtualAuthenticatorParameters/protocol}}
+    property.
+ 1. If |protocol| is null, return a [=WebDriver error=] with [=WebDriver error code=] [=invalid argument=].
+ 1. Let |transport| be the result of [=trying=] to get the |parameter|'s {{AddVirtualAuthenticatorParameters/transport}}
```
ditto

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+ 1. If |protocol| is null, return a [=WebDriver error=] with [=WebDriver error code=] [=invalid argument=].
+ 1. Let |transport| be the result of [=trying=] to get the |parameter|'s {{AddVirtualAuthenticatorParameters/transport}}
+    property.
+ 1. If |transport| is null, return a [=WebDriver error=] with [=WebDriver error code=] [=invalid argument=].
```
ditto

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+ 1. Let |transport| be the result of [=trying=] to get the |parameter|'s {{AddVirtualAuthenticatorParameters/transport}}
+    property.
+ 1. If |transport| is null, return a [=WebDriver error=] with [=WebDriver error code=] [=invalid argument=].
+ 1. Generate an arbitrary and unique string |authenticatorId|.
```
here we need to point off to wherever we declare the syntax & format requirements for |authenticatorId|

nsatragno replied:
Okay, I pointed out to the definition of `authenticatorId` above.

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+ 1. If |transport| is null, return a [=WebDriver error=] with [=WebDriver error code=] [=invalid argument=].
+ 1. Generate an arbitrary and unique string |authenticatorId|.
+ 1. Create a new [=Virtual Authenticator=] with the fields |authenticatorId|, |protocol|, |transport|,
+    |hasResidentKey|, |hasUserVerification|, and |isUserVerified|.
```
don't we need to also extract the values of |hasResidentKey|, |hasUserVerification|, and |isUserVerified| from |parameters| ?

I presume if they are not present then we let them default, yes?

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+## <dfn>Add Virtual Authenticator</dfn> ## {#sctn-automation-add-virtual-authenticator}
+
+The [=Add Virtual Authenticator=] [=extension command=] creates a software [=Virtual Authenticator=]. The
+[=extension command=] is defined as follows:
```
```suggestion
WebDriver [=extension command=] is defined as follows:
```

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+ 1. Create a new [=Virtual Authenticator=] with the fields |authenticatorId|, |protocol|, |transport|,
+    |hasResidentKey|, |hasUserVerification|, and |isUserVerified|.
+ 1. Store the authenticator in the [=Virtual Authenticator Database=].
+ 1. Return [=success=] with |authenticatorId|.
```
```suggestion
 1. Return [=success=] with data |authenticatorId|.
```

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+
+## <dfn>Remove Virtual Authenticator</dfn> ## {#sctn-automation-remove-virtual-authenticator}
+
+The [=Remove Virtual Authenticator=] [=extension command=] removes a previously created [=Virtual Authenticator=].
```
```suggestion
The [=Remove Virtual Authenticator=] WebDriver [=extension command=] removes a previously created [=Virtual Authenticator=].
```

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+
+## <dfn>Add Credential</dfn> ## {#sctn-automation-add-credential}
+
+The [=Add Credential=] [=extension command=] injects a [=Credential Key Pair=] and associated [=Authenticator Data=]
```
```suggestion
The [=Add Credential=] WebDriver [=extension command=] injects a [=Credential Key Pair=] and associated [=Authenticator Data=]
```

equalsJeffH replied:
see note below wrt "authenticator data" -- [pub key cred source](https://www.w3.org/TR/webauthn-2/#public-key-credential-source) is perhaps more approp here, and effectively includes "credential key pair"

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+
+ 1. If |parameters| is not a JSON [=Object=], return a [=WebDriver error=] with [=WebDriver error code=]
+     [=invalid argument=].
+ 1. Let |credentialId| be the result of [=trying=] to decode [=Base64url Encoding=] on the |parameter|'s
```
hm, I suppose decoding the credentialId could fail and so using "trying" here is correct, as well as for rpIdHash and privateKey below...

nsatragno replied:
Okay, I removed the "try" idiom and left the error handling specified for each parameter. What do you think?

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+ 1. Let |rpIdHash| be the result of [=trying=] to decode [=Base64url Encoding=] on the |parameter|'s
+     {{CredentialParams/rpIdHash}} property.
+ 1. If |rpIdHash| is failure, return a [=WebDriver error=] with [=WebDriver error code=] [=invalid argument=].
+ 1. If |rpIdHash| is not a valid [=rpIdHash=], return a [=WebDriver error=] with [=WebDriver error code=]
```
how does the "webdriver remote end" validate the rpIdHash ?  We ought to include the RP ID ?

nsatragno replied:
Oh, I just meant for the validation to check that the parameter is indeed a valid SHA-256 hash. I don't think it's necessary to verify that it's the actual hash of a particular RP ID.

equalsJeffH replied:
Ok, so how does one check to see if something is a valid SHA-256 hash beyond just checking the length of the byte array (32 octets) ?

nsatragno replied:
By _exactly_ doing that :P

I didn't want to "hardcode" the hash size and instead referred to the rpIdHash above :thinking:. Maybe it's better to just say, "validate that the length is 32 bytes" instead?

nsatragno replied:
On a second thought, if we are going to create a public key credential source, we need the RP ID - and maybe we can do away with passing the hash then?

nsatragno replied:
...which is what I did. PTAL.

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+     [=invalid argument=].
+ 1. Let |credentialId| be the result of [=trying=] to decode [=Base64url Encoding=] on the |parameter|'s
+     {{CredentialParams/credentialId}} property.
+ 1. If |credentialId| is failure, return a [=WebDriver error=] with [=WebDriver error code=] [=invalid argument=].
```
...however, the [definition of trying](https://www.w3.org/TR/webdriver/#dfn-try) says that returning an error is included, so this step is superfluous?  as well as the similar steps for rpIdHash and privateKey below?

nsatragno replied:
Hmmm, but we definitely want to specify the error returned. Perhaps we can expand the definition of `trying` to specify that the error returned is a WebDriver "invalid argument" error?

nsatragno replied:
I just removed the idiom altogether.

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+
+## <dfn>Get Credentials</dfn> ## {#sctn-automation-get-credentials}
+
+The [=Get Credentials=] [=extension command=] returns one {{CredentialParams}} for every [=Credential Key Pair=]
```
```suggestion
The [=Get Credentials=] WebDriver [=extension command=] returns one {{CredentialParams}} for every [=Credential Key Pair=]
```

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+        boolean hasUserVerification = false;
+        boolean isUserVerified = true;
+    };
+    </xmp>
```
I am wondering if we ought not define these webdriver-specific things with webIDL because webIDL is a single DOM-wide namespace and AFAIK this is polluting it.  Even if other specs that are clients of webdriver are using webIDL for their webdriver-ish objects/data, I'm thinking we oughta ask whether that's ok or not.  If not, we can use some other notation to describe AuthenticatorProtocol, AddVirtualAuthenticatorParameters, CredentialParams, and UserVerifiedParams.  

hm... So the webdriver spec does not use webIDL to describe POST parameters, rather it says they are a "JSON Object" and then uses tables to list the keys and value types, eg see [section 7 capabilities](https://www.w3.org/TR/webdriver/#capabilities), and examples 5, 6, 7.

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+    <xmp class="idl">
+    dictionary CredentialParams {
+        required USVString credentialId;
+        required USVString rpIdHash;
```
I'm thinking we'll want to also convey the RP ID itself if we are expecting the "remote end" to verify the `rpIdHash`.

nsatragno replied:
My idea for the verification was to just make sure it's a SHA-256 hash... not that it matches a certain RP. That way the structure accepted by `addCredential` and returned by `getCredentials` are the same, too.

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+ 1. If |privateKey| is failure, return a [=WebDriver error=] with [=WebDriver error code=] [=invalid argument=].
+ 1. If |authenticatorId| does not match any [=Virtual Authenticator=] stored in the [=Virtual Authenticator
+     Database=], return a [=WebDriver error=] with [=WebDriver error code=] [=invalid argument=].
+ 1. If |privateKey| is not a valid PKCS #8 encoded private key [[RFC5958]], return a [=WebDriver error=] with
```
```suggestion
 1. If |privateKey| is not a validly-encoded asymmetric key package containing a single private key per [[RFC5958]], return a [=WebDriver error=] with
```
Hmm, RFC5958 obsoletes PKCS \# 8, so I would not use the latter term.
Also, a `AsymmetricKeyPackage` can contain multiple private keys, hence suggesting the "single private key" stipulation.

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+     Database=], return a [=WebDriver error=] with [=WebDriver error code=] [=invalid argument=].
+ 1. If |privateKey| is not a valid PKCS #8 encoded private key [[RFC5958]], return a [=WebDriver error=] with
+     [=WebDriver error code=] [=invalid argument=].
+ 1. Create a [=Credential Key Pair=] from |privateKey| identified by |credentialId| and an [=Authenticator Data=]
```
Regarding "authenticator data", that's a structure that's [returned by authnrs](https://www.w3.org/TR/webauthn-2/#authenticator-data-perform-the-following-steps-to-generate-an-authenticator-data-structure) (in response to both create() and get()) but is not denoted as the structure authnrs use to store creds internally. I'm thinking [public key credential source](https://www.w3.org/TR/webauthn-2/#public-key-credential-source) is more correct here.

nsatragno replied:
Interesting. Done - although that means I have to specify the signature counter as a different object.

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+
+## <dfn>Virtual Authenticators</dfn> ## {#sctn-automation-virtual-authenticators}
+
+These [=extension commands=] create and interact with [=Virtual Authenticators=]: software implementations of the
```
```suggestion
These WebDriver [=extension commands=] create and interact with [=Virtual Authenticators=]: software implementations of the
```

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+:: If set to `true`, the authenticator will support [=resident credentials=].
+: |hasUserVerification|
+:: If set to `true`, the authenticator will support [=user verification=].
+: |isUserVerified|
```
I'm suspecting we need to figure out how user presence test (UP) figures in here.  E.g.. authnrs that do not support UV support UP.

nsatragno replied:
Yes, I added it above. PTAL.

equalsJeffH replied:
LGTM

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+
+ 1. If |authenticatorId| does not match any [=Virtual Authenticator=] stored in the [=Virtual Authenticator
+     Database=], return a [=WebDriver error=] with [=WebDriver error code=] [=invalid argument=].
+ 1. For every [=Credential Key Pair=], construct a corresponding {{CredentialParams}}.
```
should this be something more like:

Let |credentialsArray| be an empty array.
For each public key credential source (|credential|) managed by the authenticator identified by |authenticatorId|, construct a corresponding {{CredentialParams}} object and add it to |credentialsArray|.

..?

nsatragno replied:
Done.

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+ 1. If |authenticatorId| does not match any [=Virtual Authenticator=] stored in the [=Virtual Authenticator
+     Database=], return a [=WebDriver error=] with [=WebDriver error code=] [=invalid argument=].
+ 1. For every [=Credential Key Pair=], construct a corresponding {{CredentialParams}}.
+ 1. Return [=success=] with the array of {{CredentialParams}}.
```
```suggestion
 1. Return [=success=] with data containing |credentialsArray|.
```

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5497,267 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+
+## <dfn>Remove Credentials</dfn> ## {#sctn-automation-remove-credentials}
+
+The [=Remove Credentials=] [=extension command=] removes all [=Credential Key Pairs=] stored on a [=Virtual Authenticator=].
```
```suggestion
The [=Remove Credentials=] WebDriver [=extension command=] removes all [=Credential Key Pairs=] stored on a [=Virtual Authenticator=].
```

equalsJeffH reviewed (changes requested):
Thanks for taking this on NIna, it's a non-trivial project. Overall this is good in terms of the high-level operations and how they relate to what is web-visible by RP webapps.  there's a bunch of details to polish though, I hope this helps. there's a couple more comments I wanted to make but I'm out of time for today and this is the bulk of them. primary comments are that I don't think we ought to use webIDL, credentials managed by an auithnr should be indexed by credentialId, rather than cred key pair, will folks desire a command that removes individual creds from given authnr(?), and there's a few others below mixed into various editorial comments. 

I hope this helps, thanks again!

nsatragno committed:
Apply suggestions from code review

Commit the editorial suggestions

Co-Authored-By: =JeffH <jdhodges@google.com>

nsatragno committed:
Add missing suggestion.

nsatragno committed:
Add isUserPresent

nsatragno committed:
Constrain authenticatorId format.

nsatragno committed:
Add missing `

nsatragno committed:
Use a table instead of WebIDL for addVirtualAuthenticator.

nsatragno committed:
Move authenticatorId definition.

equalsJeffH commented on index.bs:
```diff
@@ -5565,20 +5551,83 @@ WebDriver [=extension command=] is defined as follows:
+                <td>|isUserVerified|</td>
+                <td>boolean</td>
+                <td>[TRUE], [FALSE]</td>
+                <td>[TRUE]</td>
```
Should the default for |isUserVerified| ought to match the default for |hasUserVerification| ?

nsatragno replied:
Okay, changed `|isUserVerfied|` default to `[FALSE]`.

equalsJeffH commented on index.bs:
```diff
@@ -5565,20 +5551,83 @@ WebDriver [=extension command=] is defined as follows:
+            <tr>
+                <td>|protocol|</td>
+                <td>string</td>
+                <td>`"u2f"`, `"ctap2"`</td>
```
```suggestion
                <td>`"ctap1/u2f"`, `"ctap2"`</td>
```

equalsJeffH commented on index.bs:
```diff
@@ -5520,34 +5522,18 @@ Each stored [=virtual authenticator=] has the following properties:
 ## <dfn>Add Virtual Authenticator</dfn> ## {#sctn-automation-add-virtual-authenticator}
 
 The [=Add Virtual Authenticator=] [=extension command=] creates a software [=Virtual Authenticator=]. The
 WebDriver [=extension command=] is defined as follows:
```
Upon reflection, I'd trim this down like so:
The [=Add Virtual Authenticator=] WebDriver [=extension command=] creates a software [=Virtual Authenticator=]. It is defined as follows:

nsatragno replied:
Done everywhere.

equalsJeffH commented on index.bs:
```diff
@@ -5565,20 +5551,83 @@ WebDriver [=extension command=] is defined as follows:
     </table>
 </figure>
 
+The <dfn>Authenticator Configuration</dfn> is a JSON [=Object=] containing the following parameters:
```
```suggestion
The <dfn>Authenticator Configuration</dfn> is a JSON [=Object=] passed to the [=remote end steps=] as |parameters|. It contains the following |key| and |value| pairs:
```

equalsJeffH commented on index.bs:
```diff
@@ -5565,20 +5551,83 @@ WebDriver [=extension command=] is defined as follows:
 The [=remote end steps=] are:
 
  1. If |parameters| is not a JSON [=Object=], return a [=WebDriver error=] with [=WebDriver error code=]
      [=invalid argument=].
```
suggest add this after above line:

Note: |parameters| is a [=authenticator configuration=] object.

nsatragno replied:
Done.

equalsJeffH commented on index.bs:
```diff
@@ -5565,20 +5551,83 @@ WebDriver [=extension command=] is defined as follows:
+ 1. Let |authenticator| be a new [=Virtual Authenticator=].
+ 1. For each enumerable [=own property=] in |parameters|:
+    1. Let |key| be the name of the property.
+    1. Let |value| be the result of [=getting a property=] named |name| from |parameter|.
```
```suggestion
    1. Let |value| be the result of [=getting a property=] named |key| from |parameter|.
```
...?

nsatragno replied:
Done.

equalsJeffH commented on index.bs:
```diff
@@ -5565,20 +5551,83 @@ WebDriver [=extension command=] is defined as follows:
+ 1. For each enumerable [=own property=] in |parameters|:
+    1. Let |key| be the name of the property.
+    1. Let |value| be the result of [=getting a property=] named |name| from |parameter|.
+    1. If there is no matching `key` for |key| in the [=Authenticator Configuration=], return a [=WebDriver error=] with
```
s/ in the [=Authenticator Configuration=] / in |parameters| /

nsatragno replied:
Done.

equalsJeffH commented on index.bs:
```diff
@@ -5565,20 +5551,83 @@ WebDriver [=extension command=] is defined as follows:
+ 1. For each property in [=Authenticator Configuration=]:
+    1. If `key` is not a defined property of |authenticator|, return a [=WebDriver error=] with [=WebDriver error code=]
+        [=invalid argument=].
+ 1. Generate an unique string |authenticatorId| out of Unreserved Characters [[RFC3986]].
```
s/ out of Unreserved Characters  [[RFC3986]]. / using characters from the `unreserved` production defined in [Appendix A](https://tools.ietf.org/html/rfc3986#appendix-A) of [[RFC3986]]. /

nsatragno replied:
I missed this in the refactor - moved this text to the definition of authenticatorId.

equalsJeffH commented on index.bs:
```diff
@@ -5565,20 +5551,83 @@ WebDriver [=extension command=] is defined as follows:
+        [=invalid argument=].
+ 1. Generate an unique string |authenticatorId| out of Unreserved Characters [[RFC3986]].
+    <div class="note">
+        Note: This string can be a UUID [[RFC4122]].
```
s/ This string can be a UUID / [=generating a uuid|Generating this string as a UUID=] is recommended /

nsatragno replied:
Moved this above, too.

equalsJeffH reviewed (commented):
thx @nsatragno, this LGTM overall, some modest suggestions below....

nsatragno committed:
Trim down definitions.

nsatragno committed:
Apply suggestions from code review

Apply trivial changes

Co-Authored-By: =JeffH <jdhodges@google.com>

nsatragno committed:
Fix up addVirtualAuthenticator.

nsatragno committed:
Fix up AddCredential.

nsatragno committed:
Several misc fixes.

equalsJeffH commented on index.bs:
```diff
@@ -5516,7 +5516,7 @@ Each stored [=virtual authenticator=] has the following properties:
     </div>
 : |protocol|
-:: The protocol the [=Virtual Authenticator=] uses to encode its messages. Either `ctap2` for the [[FIDO-CTAP]] format
+:: The protocol the [=Virtual Authenticator=] speaks: either `ctap2` or `ctap1-u2f` [[FIDO-CTAP]].
```
we probably should s/ `ctap1-u2f` / `ctap1/u2f` /

nsatragno replied:
Done!

nsatragno reviewed (commented):
> will folks desire a command that removes individual creds from given authnr

I can see how this would be useful... I should probably add it.

nsatragno committed:
Fix typo.

nsatragno committed:
Add removeCredential and rename removeCredentials

nsatragno commented:
I added `removeCredential` and renamed `removeCredentials` to `removeAllCredentials`. I think you can give this another look.

nsatragno committed:
touch up authenticatorId definition.

nsatragno committed:
Address comments.

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5504,379 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+:: An non-null string made using up to 48 characters from the `unreserved` production defined in Appendix A of [[RFC3986]]
+    that uniquely identifies the [=Virtual Authenticator=].
+
+    Note: [=generating a uuid|Generating this string as a UUID=] is recommended.
```
I'm having 2nd thoughts about this unqualified recommendation of gen'g a UUID for authnrID -- it is seeming like overkill to me now.  In an isolated environment one could just number the authnrs 1, 2, 3, etc, yes?  plus different "sessions" are disambiguated by `session id`, so they won't have authnrID collisions.  what do you think?

nsatragno replied:
Sounds reasonable. Removed this note and reference.

equalsJeffH commented on index.bs:
```diff
@@ -5483,6 +5504,379 @@ At this time, one [=credential property=] is defined: the [=resident key credent
+ 1. Let |credentialId| be the result of decoding [=Base64url Encoding=] on the |parameters|' |credentialId| property.
+ 1. If |credentialId| is failure, return a [=WebDriver error=] with [=WebDriver error code=] [=invalid argument=].
+ 1. Let |rpId| be the the |parameters|' |rpId| property.
+ 1. If |rpId| is not a valid [=RP ID=], return a [=WebDriver error=] with [=WebDriver error code=] [=invalid argument=].
```
Hm, I spose havin [the [=RP ID=] autolink](https://www.w3.org/TR/webauthn/#rp-id) gives the reader enough context in order to figure out how to validate an alleged RP ID value?

nsatragno replied:
That is my assumption, yes :)

equalsJeffH commented:
> I added `removeCredential` and renamed `removeCredentials` to `removeAllCredentials`. I think you can give this another look.

super, LGTM -- are we ready to open a webauthn PR?  I also made a couple further minor comments just now.

nsatragno committed:
Remove UUID recommendation.

nsatragno committed:
Change isUserPresent to isUserConsenting.

nsatragno committed:
Add description to credentialParameters table.

Also fix virtual authenticator `credential` still being refered
as key pair instead of public key credential source.

nsatragno committed:
Encode private key using COSE instead.

nsatragno committed:
Add userHandle.

nsatragno committed:
nit.

nsatragno committed:
nit

nsatragno committed:
Revert "Encode private key using COSE instead."

This reverts commit cf614d48b38f620b4f3775723178457702eededb.

nsatragno committed:
Add |isResidentCredential| to AddCredential.

nsatragno committed:
Add check for resident credential.

nsatragno committed:
Constrain the pk to ECDSA P-256

nsatragno committed:
Return an error if userHandle is null and is resident key.
